File: src/main/scala/ml/parameters.scala
package ml

import java.lang.reflect.Modifier

import scala.collection.mutable

/**
 * Describes a parameter key.
 * @param parent parent id
 * @param name parameter name
 * @param doc documentation
 * @param defaultValue default value
 * @tparam T parameter value type
 */
class Param[+T](
    val parent: Identifiable,
    val name: String,
    val doc: String,
    val defaultValue: Option[T]) extends Serializable {

  override def toString: String = s"$name: $doc (default: $defaultValue)"
}

/**
 * Trait for components with parameters.
 */
trait Params {

  /** Returns all parameters. */
  def params: Array[Param[_]] = {
    val methods = this.getClass.getMethods
    methods.filter { m =>
      Modifier.isPublic(m.getModifiers) &&
        classOf[Param[_]].isAssignableFrom(m.getReturnType) &&
        m.getParameterTypes.isEmpty
    }.map(m => m.invoke(this).asInstanceOf[Param[_]])
  }

  /** Gets a param by its name. */
  def getParam(paramName: String): Param[_] = {
    val m = this.getClass.getMethod(paramName, null)
    assert(Modifier.isPublic(m.getModifiers) &&
      classOf[Param[_]].isAssignableFrom(m.getReturnType))
    m.invoke(this).asInstanceOf[Param[_]]
  }

  /** Validate parameters specified by the input parameter map. */
  def validateParams(paramMap: ParamMap): Unit = {}
}

object Params {
  val empty: Params = new Params {
    override def params: Array[Param[_]] = Array.empty
  }
}

/**
 * A param -> value map.
 */
class ParamMap private (params: mutable.Map[Identifiable, mutable.Map[String, Any]]) extends Serializable {

  def this() = this(mutable.Map.empty[Identifiable, mutable.Map[String, Any]])

  /**
   * Puts a new (param, value) pair.
   */
  def put[T](param: Param[T], value: T): this.type = {
    if (!params.contains(param.parent)) {
      params(param.parent) = mutable.HashMap.empty[String, Any]
    }
    params(param.parent) += param.name -> value
    this
  }

  /**
   * Gets the value of the input parameter or the default value if it doesn't exist.
   */
  def getOrDefault[T](param: Param[T]): T = {
    val value = params.get(param.parent) match {
      case Some(map) =>
        map.getOrElse(param.name, param.defaultValue)
      case None =>
        param.defaultValue
    }
    value.asInstanceOf[T]
  }

  /**
   * Filter this parameter map for the given parent.
   */
  def filter(parent: Identifiable): ParamMap = {
    val map = params.getOrElse(parent, mutable.Map.empty[String, Any])
    new ParamMap(mutable.Map(parent -> map))
  }

  /**
   * Make a deep copy of this parameter set.
   */
  def copy: ParamMap = new ParamMap(params.clone())

  override def toString: String = {
    params.map { case (parentId, map) =>
      s"{$parentId: $map}"
    }.mkString("{", ",", "}")
  }
}

File: src/main/python/ml/parameters.py
class Param:

    def __init__(self, parent, name, doc, defaultValue):
        self.parent = parent
        self.name = name
        self.doc = doc
        self.defaultValue = defaultValue

    def withValue(self, value):
        return ParamPair(self, value)

    def __str__(self):
        return "{0}/{1}: {2} (default: {3})".format(self.parent, self.name, self.doc, self.defaultValue)

    def __repr__(self):
        return "{0}/{1}".format(self.parent, self.name)

class ParamPair:

    def __init__(self, param, value):
        assert isinstance(param, Param)
        self.param = param
        self.value = value

class ParamMap:

    def __init__(self):
        self.params = {}

    def put(self, param, value):
        self.params[param] = value
        return self

    def getOrDefault(self, param):
        return self.params[param] if param in self.params else param.defaultValue

    def copy(self):
        newMap = ParamMap()
        newMap.params = self.params.copy()
        return newMap

    def __repr__(self):
        return self.params.__repr__()

class ParamGridBuilder:

    def __init__(self):
        self.paramGrid = {}

    def add(self, param, value):
        return self.addMulti(param, [value,])

    def addMulti(self, param, values):
        self.paramGrid[param] = values
        return self

    def build(self):
        paramMaps = [ParamMap(),]
        for (param, values) in self.paramGrid.items():
          newParamMaps = []
          for paramMap in paramMaps:
              for v in values:
                  newParamMap = paramMap.copy()
                  newParamMap.put(param, v)
                  newParamMaps.append(newParamMap)
          paramMaps = newParamMaps
        return paramMaps
